pkg/cmd/run.go

```diff
@@ -144,30 +163,70 @@ func NewRunCommand(factory factory.Factory, streams genericclioptions.IOStreams)
 
 // Validate validates the arguments and flags populating RunOptions accordingly.
 func (o *RunOptions) Validate(cmd *cobra.Command, args []string) error {
+	// Selector can only be used in conjunction with tracer.
+	o.tracerDefined = cmd.Flag("tracer").Changed
+	if !o.tracerDefined && cmd.Flag("selector").Changed {
+		return fmt.Errorf(tracerNeededForSelectorErrString)
+	}
+
+	switch o.tracer {
+	case bpftrace, bcc:
+	default:
+		return fmt.Errorf(tracerNotFound, o.tracer)
+	}
+
+	parsed, err := tracejob.NewSelector(o.selector)
+	if err != nil {
+		return fmt.Errorf(err.Error())
+	}
+	o.parsedSelector = parsed
+
+	// All filtering must be via selector if tracer is specified.
 	containerFlagDefined := cmd.Flag("container").Changed
-	switch len(args) {
-	case 1:
-		o.resourceArg = args[0]
-		break
-	// 2nd argument interpreted as container when provided
-	case 2:
-		o.resourceArg = args[0]
-		o.container = args[1]
-		if containerFlagDefined {
-			return fmt.Errorf(containerAsArgOrFlagErrString)
+	if o.tracerDefined && containerFlagDefined {
+		return fmt.Errorf(containerAsSelectorErrString)
+	}
+	if o.tracerDefined && len(args) > 0 {
+		return fmt.Errorf(resourceAsSelectorErrString)
+	}
+
+	if !o.tracerDefined {
```

Comment:
One thing that confused me when first using/modifying kubectl-trace was the order in which it parsed things, e.g. when it would magically interpret positional arguments as Pods vs. containers vs. general arguments - would you mind adding some docs here?

Reply:
Thanks, I can most certainly do that. Based on other review feedback if we decide to keep old flags then I'll add a comment here otherwise if we do a breaking change then I'll rewrite the full flag documentation anyways.